From 19858bfbded34cb9923d67d572c83244d2e567d7 Mon Sep 17 00:00:00 2001
From: Lon Hohberger <lon@users.sourceforge.net>
Date: Wed, 3 Aug 2011 10:21:44 -0400
Subject: [PATCH] Fix incorrect return value on hash mismatch

Resolves: rhbz720767

Signed-off-by: Lon Hohberger <lon@users.sourceforge.net>
---
 client/mcast.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/client/mcast.c b/client/mcast.c
index 4aaa4e4..4fbb422 100644
--- a/client/mcast.c
+++ b/client/mcast.c
@@ -137,14 +137,14 @@ tcp_exchange(int fd, fence_auth_type_t auth, void *key,
 	if (tcp_challenge(fd, auth, key, key_len, timeout) <= 0) {
 		/* Challenge failed */
 		printf("Invalid response to challenge\n");
-		return 0;
+		return 1;
 	}
 
 	/* Now they'll send us one, so we need to respond here */
 	dbg_printf(3, "Responding to TCP challenge\n");
 	if (tcp_response(fd, auth, key, key_len, timeout) <= 0) {
 		printf("Invalid response to challenge\n");
-		return 0;
+		return 1;
 	}
 
 	dbg_printf(2, "TCP Exchange + Authentication done... \n");
-- 
1.7.3.4

